lte/gateway/c/core/oai/tasks/s1ap/s1ap_state_manager.cpp

```diff
@@ -119,11 +123,17 @@ void S1apStateManager::free_state() {
   if (hashtable_ts_destroy(&state_cache_p->mmeid2associd) != HASH_TABLE_OK) {
     OAI_FPRINTF_ERR("An error occurred while destroying assoc_id hash table");
   }
+  free_wrapper(reinterpret_cast<void**>(&state_cache_p));
```

Comment:
would this work the same using C explicit cast  (we do it in other cpp files)
```
free_wrapper((void**) &state_cache_p);
```
or just use static case?

Reply:
Oh, I changed from the explicit cast to use the reinterpret_cast<T> on suggestion from cpplint (mostly for readability?):

https://github.com/magma/magma/commit/b36270e94bdc29e417c0a7bb5ef9cf4f44259bf8#diff-1f2fe1fb53fd6b8618167066d82498d0712d2b060c9251c7020a528af033b900R125

Do we prefer the c-style cast?

Reply:
no, let's follow cpplint suggestion.

Reply:
Oh I think @uri200 is suggesting to use static_cast?

Reply:
yep, per what I could read static_cast is the recommended one for cpp. The c style is what we have on other cpp files

Reply:
If static_cast works, that's what you should reach for first. I'm unable to predict whether it will work here.

Reply:
This particular instance does not work:

```
/home/vagrant/magma/lte/gateway/c/core/oai/tasks/s1ap/s1ap_state_manager.cpp: In function ‘void magma::lte::free_s1ap_state(s1ap_state_t*)’:
/home/vagrant/magma/lte/gateway/c/core/oai/tasks/s1ap/s1ap_state_manager.cpp:126:50: error: invalid static_cast from type ‘s1ap_state_t**’ {aka ‘s1ap_state_s**’} to type ‘void**’
  126 |   free_wrapper(static_cast<void**>(&state_cache_p));
      |                                                  ^
```

We are able to use static_cast above in line 68 though; I'll push an update

Reply:
is it due to double pointer?

Reply:
it's due to the semantic of `void*` (pointer to anything) and `void**` (pointer to untyped pointer). `static_cast` can't convert to/from `void**` because it doesn't (can't?) know what it is. `reinterpret_cast` appears to be the right answer for `void**`

https://stackoverflow.com/questions/26277391/why-cant-static-cast-a-double-void-pointer